### ISSUE TITLE
Create compliance definition and script for identifying non-compliant sites in Google Sheets crawl data

### ISSUE THREAD

**Issue by SebastianZimmeck** [privacy-tech-lab/gpc-web-crawler#136]
@katehausladen provided some initial analysis accuracy analysis as shown in [our draft paper](https://drive.google.com/open?id=1lkE6BdyVFfmE2fdPvDWKkvLcr6Rk8wqV&usp=drive_fs) (section 3.5). Starting with the September crawl (#118) we should come up with a protocol to check for each crawl going forward 100 randomly selected sites manually whether the crawl results are accurate. As we are crawling over longer periods of time we might otherwise see a drift in accuracy, for example, due to code changes or site changes, and, thus, should keep an eye on it. 

I am particularly concerned about the following:

- uspapi_before_gpc, uspapi_after_gpc
- usp_cookies_before_gpc, usp_cookies_after_gpc
- OptanonConsent_before_gpc, OptanonConsent_after_gpc
- gpp_before_gpc, gpp_after_gpc
- usps_before_gpc, usps_after_gpc 
- USPS implementation
- error
- Well-known

A few comments:

- @katehausladen came up with strategy to check the ground truth while the crawl was running. That was useful because site loads may differ in terms of which ad networks and other third party a site loads in one run vs another run. Now, if we do a post-hoc analysis after the crawl, we cannot do that as the crawl has already happened. But I think that it is OK because I am less concerned about the `urlClassification` results but more about the above. Question: do the above change from load to load? I do not think that should be the case since, for example, a site should set the OptanonConsent cookie on every load. But we need to check that.
- Also, how do we ensure that our random selection covers all the cases? Having one instance of the GPP implementation behaving correctly is not meaningful. We should aim for around ten instances per condition. So, how do we select random sites but ensuring sufficient instances? Our set of sites is skewed. For example, most sites do not have a GPP string. So, total random selection may not lead to sufficient coverage. Maybe, randomly select from the set of sites that have a GPP implementation per the crawl and confirm 10 positive instances and randomly select from the set of non-GPP sites 10 to confirm negative instances. This analysis is further complicated as there are different sub-conditions, for example, USP API opts out after receiving a GPC signal vs USP API already opted out before receiving a GPC signal. In the first case we have an instance of a string change that our crawl needs to get accurate.

The bottom line is, we need a protocol that allows us to check the analysis accuracy of our different conditions (including sub-conditions) for every crawl to keep track of analysis accuracy over time. Since we need to do it every crawl and it involves manual work, it should be manageable time-wise but also meaningful.

@natelevinson10 will take the lead here and work with @franciscawijaya and @eakubilo before starting the next crawl.

**Comment by natelevinson10** [comment]
Did a quick review of the manual data we collected a couple of weeks ago and targeted instances of a mismatch (marked in red) signaling our ground truth was different than the crawl data. I used several VPN locations (California, multiple Colorado, Virginia, and no VPN (CT)) and gave ample time to let all of the site content to load. 

I was not able to find a single instance of our manual data changing from what we had reported, except for bumble.com 's USPapi_before being "1YNN" instead of the reported "1YYN", and I would chalk this up to a manual error on our end. It would seem that for the mismatches of data from crawl to manual, the manual data is more accurate.

As for the comments left by @SebastianZimmeck above, cookies do seem to load on every refresh - I have yet to find an instance where an OptAnon cookie does not load where it should. I plan to do some more testing on this to be certain over the next few days. As for our site sample skew, I believe it could be worth it to have a subset of websites we know to have GPP / OTGPPConsent data. A thought I have is to compile a list of websites we know to have all our needed behaviors (i.e. USP API opts out after receiving a GPC signal vs USP API already opted out before receiving a GPC signal ETC.) as these are crucial in our crawl list to get a holistic representation of results. I plan on seeing if there is a list of directory of websites having certain attributes that could simplify a search for these websites, if it is something we choose to do.

**Comment by SebastianZimmeck** [comment]
Thanks, @natelevinson10!

**Comment by SebastianZimmeck** [comment]
As discussed today, @franciscawijaya and @natelevinson10 will come up with a protocol of selecting 100 sites for a manual spotcheck of the first batch that has sufficient coverage (say, at least 5 positive instances, if possible) for each item we test for.

@franciscawijaya and @natelevinson10, you can write the protocol here in the issue for the time being.

**Comment by natelevinson10** [comment]
To accurately assess the accuracy of the crawl data across the crawl as a whole, our protocol should focus on selecting a representative and stratified sample of 100 sites for manual review. The plan is to run a crawl batch and select 100 sites to review via the constraits below. We will then compare our results from the crawl to our manual review.
 
- Ensure 15 instances of usp api and cookies, and optanon data points changing from before to after. This accounts for 30% of our site list, and will give us ample information on false positives/negatives as these are the most common instances of data.
Ensure 15 instances of gpp before/after. In our CO crawl, we only had 3 instances, that is not enough to draw any conclusion about accuracy.
- Ensure 15 instances of OTGPPConsent before/after. Same as above, but only 2 instances.
- Next 40 sites will be selected randomly from alternating 100s places (i.e. first 10 from 0-200, next 10 from 201-400, next 10 from 401-600) and so on. This will make sure we pull sites from all points of progression in the crawl.
- The final 15 sites will sites flagged with human check errors. This is to address our "edge cases" and confirm that sites flagged with human check errors are not slipping through any accounting for any data mismatches.

After compiling this list of 100 sites, we will manually check them in a similar fashion as we did with the CO crawl [here](https://docs.google.com/spreadsheets/d/1jRp9exj2Sc5ShaMijdLObWQ0Q-tJ-lck/edit?gid=670930703#gid=670930703). We will use our same methodology for verifying the maunal results [here](https://docs.google.com/spreadsheets/d/1jRp9exj2Sc5ShaMijdLObWQ0Q-tJ-lck/edit?gid=1551638211#gid=1551638211). Here is our initial plan @franciscawijaya and I reviewed, let us know what you think.

**Comment by Mattm27** [comment]
The methodology outlined above provides a strong approach by ensuring a randomized stratified sample that is well-representative of the entire crawl. The constraints effectively balance random selection with targeted verification of key conditions, making the accuracy check both meaningful and manageable. However, it may be beneficial to update the methodology slightly depending on the output and what the complete crawl data looks like. For example, if OTGPPConsent appears in only a very small portion of the data, we may decide to ignore manual checking for those instances to focus on more prevalent conditions.

>  - Ensure 15 instances of usp api and cookies, and optanon data points changing from before to after. This accounts for 30% of our site list, and will give us ample information on false positives/negatives as these are the most common instances of data.
>  - Ensure 15 instances of gpp before/after. In our CO crawl, we only had 3 instances, that is not enough to draw any conclusion about accuracy.
>  - Ensure 15 instances of OTGPPConsent before/after. Same as above, but only 2 instances.
>  - Next 40 sites will be selected randomly from alternating 100s places (i.e. first 10 from 0-200, next 10 from 201-400, 
next 10 from 401-600) and so on. This will make sure we pull sites from all points of progression in the crawl.
>  - The final 15 sites will sites flagged with human check errors. This is to address our "edge cases" and confirm that sites flagged with human check errors are not slipping through any accounting for any data mismatches. 

A sample size of just 10 sites that were discussed in last week's meeting would not be sufficient to draw meaningful conclusions about the accuracy, as it wouldn’t capture enough variability across different conditions. However, we could explore whether a slightly reduced sample—around 50 - 75 sites instead of 100—would still provide reliable insights while making the manual review process more efficient. This could be done by ensuring coverage of key conditions while slightly reducing the number of random selections.

Since we haven’t run a full crawl since the fall, moving forward with this methodology is the best way to verify the accuracy of the crawl results currently being collected by @samir-cerrato. Running this manual check now will help us confirm whether a drift in accuracy occurs in the future while providing a baseline for future comparisons. The manual check site list will not remain constant across months but will instead change based on the output data of each crawl, ensuring that we capture meaningful variations in the data.

We have laid out a systematic way to execute these manual checks from our CO crawl to ensure consistency in our approach [here](https://docs.google.com/spreadsheets/d/1jRp9exj2Sc5ShaMijdLObWQ0Q-tJ-lck/edit?gid=1551638211#gid=1551638211). To help track discrepancies over time, I can put together a new sheet in the GPC folder that will log discrepancies found each month and allow us to monitor how they fluctuate or change over time.

**Comment by SebastianZimmeck** [comment]
@Mattm27 created a [Google Sheet to keep track of the performance results over time](https://docs.google.com/spreadsheets/d/1ZNE0hywsv-rVlcgsFpEgwta5d2Mmx5ik4DWxAvTaNFM/edit?usp=sharing). I changed the naming convention of the tabs to align them with our crawl tabs naming convention and make them more clear.

@Mattm27, can you move ahead in your Google Sheet with the analysis for the [February 2025 CO crawl](https://docs.google.com/spreadsheets/d/1jd1hTVkc3c8Nfrs40UTK0RXPbXvpnNahVjNhIo6lLT8/edit?usp=sharing)?

**Comment by Mattm27** [comment]
To efficiently choose sites to review after each crawl manually, I put together a [Google Apps Script](https://script.google.com/u/0/home/projects/1d3Y6KMDDV3pq_sTNGundnJ7OnzkJTq6CVJFpPvNR6EaUpPAXSSeZvvOz). This script randomly selects a sample of websites from our Google Sheet containing web crawler data for manual review. The script focuses on specific columns of interest (uspapi, usp cookies, OptanonConsent, gpp, usps, Well-known, gpp_version, etc...) and checks if a row has valid non-null data in at least one of these columns. For each column of interest, 5 random rows with valid data are selected and added to an output sheet titled "Accuracy Review" for manual ground truth analysis. The main functions of the script include:

- `selectRandomSites`: Collects and writes the random sample to the "Manual Review" sheet.
- `getRandomSamples`: Selects random rows from an array.
- `rowHasValidData`: Ensures a row contains valid data in specified columns.

I am still in the process of tweaking the parameters and output of the script. Still, this script should help us efficiently assess the accuracy of crawler data by manually verifying a representative sample of sites. Once the script is complete, I will begin to conduct an accuracy test for the CO Febuary 2025 crawl.

**Comment by Mattm27** [comment]
In order to automate the detection of non-compliant sites, we will evaluate three key conditions after a GPC signal is sent.
 
1. The third character of `US Privacy Strings` (uspapi and usp cookie) should be "Y" if these values are set initially, indicating that the site acknowledged the opt-out request; if they were null to begin with, compliance is not affected.
2. If `OptanonConsent_before_gpc` contains "isEnabled=0", then `OptanonConsent_after_gpc` must be "isEnabled=1" after GPC, confirming that the site properly enabled opt-out preferences.
3. The `decoded_gpp_after_gpc` field must indicate opt-out values set to 1 in relevant sections—specifically, in the U.S. National section (usnatv1), SaleOptOut, SharingOptOut, and TargetedAdvertisingOptOut should all be 1. 

If a site fails any of these checks and has valid data for the respective fields, it is flagged as potentially non-compliant and added to a new sheet for further review.

**Comment by SebastianZimmeck** [comment]
Just for clarity, the [finding of non-compliant sites](https://github.com/privacy-tech-lab/gpc-web-crawler/issues/136#issuecomment-2704675161) is actually slightly different from the rest of this issue, which is the checking of the accuracy.

**Comment by franciscawijaya** [comment]
Having established the methodology and ensured our understanding is consistent for the execution of the accuracy check, Matt and I will be doing the accuracy check for CT and CA crawl this week and hopefully we will have the results by the next meeting. After that is done, I will also ensure that the google sheet is [clear](https://github.com/privacy-tech-lab/gpc-web-crawler/issues/143).

**Comment by Mattm27** [comment]
After several attempts to conduct the ground truth analysis for our accuracy reviews for each crawl, it has become increasingly difficult to use the Selenium grid to view the crawl browser and directly check the data the crawler is gathering. The interface is slow and does not allow copying and pasting commands into the console to pull specific values (such as USPAPI and GPP). Additionally, the Selenium grid randomly resets between different sites. While the crawler is collecting data correctly, validating this data—particularly the USPS and GPP data—within the Selenium grid interface while the crawler is running proves challenging. It would be more efficient to conduct the accuracy review manually using a fresh Firefox instance without the crawler running and then compare the data we gathered with the data collected by the crawler. Though I have been able to validate a few sites using our original ground truth methodology, switching to this method may be a more scalable alternative, enabling us to arrive at the same conclusions. I will also explore other alternative methods to improve the process.

**Comment by eakubilo** [comment]
If Selenium Grid's noVNC (a way to view desktop environments in your browser) isn't featureful enough (no copy and paste), what you can do is download a dedicated VNC client which will allow you to copy and paste from your local machine to the desktop environment. To do this, you need to expose the VNC port in `compose.yml`. You do this by adding "5900:5900" to the list of ports. This should then allow you to use a VNC client like TigerVNC or TightVNC. I haven't tried either of them out yet, but this should be the fastest path forward. To reiterate, what I suggest you to do is:
1. Download a dedicated VNC client like TigerVNC or TightVNC
2. Add the port mapping "5900:5900" in the list of port mappings in `compose.yml`
3. Start the crawler as you intend to for the crawl accuracy check and perform your methodology through the VNC checker.

If that doesn't work, I think the next thing to do is to add buttons to the extension that propagate privacy strings to the rest API. This shouldn't be hard - just add buttons to the modal and add an extra endpoint to the rest API. Then, you can perform your accuracy check by navigating to the websites, and just clicking the buttons on the extension. 

Either option should work to preserve the methodology as you described.

**Comment by Mattm27** [comment]
I was successfully able to download TigerVNC and use that to view the crawler browser instead of Selenium Grid's noVNC. On my mac, I am still not able to copy and paste commands into the browser console using TigerVNC, however the interface runs much smoother. With this new dedicated VNC client, our original methodology for accuracy reviews seems like it will still be feasible as long as we set long enough timeouts for ourselves to collect all of the necessary data. @eakubilo, thank you for providing this solution!

**Comment by franciscawijaya** [comment]
I ran into some problems the past week while attempting to do the accuracy check on my local device. While I have yet to finish the CO accuracy check, I was finally able to troubleshoot all of the problems that I encountered. I have thus taken the liberty to [update the wiki to include a detailed documentation](https://github.com/privacy-tech-lab/gpc-web-crawler/wiki/Instructions-for-Lab-Members-Performing-Crawls#accuracy-check-protocol) for the three-pronged process of the accuracy test methodology: random selection of sample sites, setting up VNC, and accuracy check crawler methodology. Now that I have successfully debugged the problems, I would be able to finish the CO accuracy check by the next week's meeting.

**Comment by franciscawijaya** [comment]
One thing that I was discussing with Matt was to push the changes of adding the port 5900:5900 to the main branch just so that we don't have to manually add and remove that line whenever we want to run the accuracy check test. I think this could be a good idea as adding the line in the code does not necessarily means that the VNC would automatically start. Given that the VNC environment depends on whether we have this line `environment: - SE_START_VNC=false` and access TigerVNC app or not, it might be helpful to just have the port 5900:5900 be in the main branch of the codebase.

**Comment by Mattm27** [comment]
@katehausladen, we completed our first accuracy check for the [CT Feb 2025 Crawl Data](https://docs.google.com/spreadsheets/d/1ZNE0hywsv-rVlcgsFpEgwta5d2Mmx5ik4DWxAvTaNFM/edit?gid=1030654467#gid=1030654467) using a similar ground truth methodology used in your thesis paper. During the process, we identified a few discrepancies in the crawl data. Specifically, privacy strings that were present on the website were returned as null by the crawler. In each case, this occurred before the GPC signal was sent. We suspect this may be due to the crawler not having enough time to fully analyze the website before the privacy strings were populated.

We wanted to check with you to see if you've come across any similar discrepancies while accuracy-checking your data for your thesis. Let us know if you have any thoughts or experiences with this!

**Comment by katehausladen** [comment]
This was not an issue for me--the main accuracy issue for me was variability between site loads. You could try increasing the crawler timeouts and see if that helps. You could also see if these sites are related in any way and therefore have similar functionality. I know there was some issue where sites blocked script injection but if it's only impacting before gpc, I don't think this is the problem.

**Comment by franciscawijaya** [comment]
I finished the Colorado Ground Truth Accuracy Check and discovered some high-level points during the process that I would love to bring up during the meeting today: 
- The Google AppScript seems to generate a batch of sites that contain multiple duplicates (discussed this with Matt and we can look into refinements to the script on the randomization function)
- Due to the slower interface of the VNC, it was generally tougher to manually check the data before the GPC signal is sent as compared to our previous methodology prior to dockerization (for future accuracy checks, we should probably look into ways to increase timeout for before GPC signal is sent so not just the timeout for every site load after gpc signal is sent) 

I would also like to confirm and check on some other details during the call: 
1)  While the accuracy check comparing the manual data collection and the crawl data collected simultaneous to the manual check is generally consistent, I also tried to compare the manual data collection on this accuracy check run to Samir's crawl data and discovered some discrepancies. Although they might vary due to time gap/IP addresses/browser session/etc, I think it might be worth pointing these discrepancies out to make sure our accuracy check methodology is airtight: 
- different gpp strings between overall crawl and accuracy check crawl 
- gpp string during overall crawl vs empty gpp string during accuracy check crawl 
- empty gpp string during overall crawl vs gpp string during accuracy check crawl

2) Addition of error phrases
I think manual accuracy check might be a good opportunity to identify error phrasing that might not be in our general error checks pool. I found an error sentence that is not flagged as an error by our crawl (ie. HumanCheck, etc): Error 1000 "DNS points to prohibited IP" which should come under HumanCheckError

3) The existence duplicates actually allowed me to identify a possible inquiry. While I started with doing checks in batches of 5 to minimize human checks. I eventually also did a check with an immediate batch of 30 and realized for some of sites that are duplicates, the crawl did not capture data for their second run. I'm not sure if this would be a problem since generally our crawl  does not concern itself with analyzing duplicates, but rather only a redo which is done subsequently (different from how the duplicates ordering by the appscripts for the accuracy check).

**Comment by franciscawijaya** [comment]
> * The Google AppScript seems to generate a batch of sites that contain multiple duplicates (discussed this with Matt and we can look into refinements to the script on the randomization function)

Action: 
- @Mattm27 and. I would look into the script to see the duplicate problem so as to prevent such a problem to reoccur 
- Update the accuracy check methodology in an alternating way.
For example, we start with doing an accuracy check for just Connecticut in January, and then after a few months, we can do the accuracy check just for Colorado and then the next one we would do in the next few months is California and we keep rotating that way. 

> * Due to the slower interface of the VNC, it was generally tougher to manually check the data before the GPC signal is sent as compared to our previous methodology prior to dockerization (for future accuracy checks, we should probably look into ways to increase timeout for before GPC signal is sent so not just the timeout for every site load after gpc signal is sent)

Action: 
Look into ways to increase timeout for the site load before GPC signal is sent 

> * different gpp strings between overall crawl and accuracy check crawl
> * gpp string during overall crawl vs empty gpp string during accuracy check crawl
> * empty gpp string during overall crawl vs gpp string during accuracy check crawl

This is not a big problem given the variables involved: time gaps, different timeouts and site loads

> 2\. Addition of error phrases
>     I think manual accuracy check might be a good opportunity to identify error phrasing that might not be in our general error checks pool. I found an error sentence that is not flagged as an error by our crawl (ie. HumanCheck, etc): Error 1000 "DNS points to prohibited IP" which should come under HumanCheckError

Action: 
- Add this particular error line in the codebase. I'm hoping I can finally read through the entire codebase to share my thoughts for the documentation with @eakubilo and would include this error line while doing this. 
- Add this as an additional point in the documentation for the accuracy check.

> 3\. The existence of duplicates actually allowed me to identify a possible inquiry. While I started with doing checks in batches of 5 to minimize human checks, I eventually also did a check with an immediate batch of 30 and realized for some of sites that are duplicates, the crawl did not capture data for their second run. I'm not sure if this would be a problem since generally our crawl  does not concern itself with analyzing duplicates, but rather only a redo which is done subsequently (different from how the duplicates ordering by the appscripts for the accuracy check).

Should be resolved when the duplicates are removed

**Comment by franciscawijaya** [comment]
Updated and finalized the wiki on the methodology for the accuracy check! Feel free to let me know what your thoughts and if there's anything unclear @samir-cerrato

**Comment by avan36** [comment]
Matt and I met today (Wednesday), we have an understanding of how to parse the GPP strings. For detecting non compliant sites, Matt has already set up good USP API, USP cookie, and OptanonConsent checks. We are going to migrate the code to a new google colab to align with the rest of the work the lab has done and also so we can use python to unlock more functionality for parsing the GPP string.

**Comment by SebastianZimmeck** [comment]
> ... so we can use python to unlock more functionality for parsing the GPP string.

Great! Generally, what would be helpful is if you keep an eye on extending your Python approach towards a full dashboard of visual results presentation and filtering. We do not need to do this now. But it would be good to have the option in the future. Imagine a regulator can easily filter according to specific criteria for non-compliant sites and visualize the returned results.

**Comment by SebastianZimmeck** [comment]
Going forward, let us use this issue for the script for identifying non-compliant sites in the Google Sheets only. For designing and testing the protocol for determining crawl accuracy over time I created a [separate issue](https://github.com/privacy-tech-lab/gpc-web-crawler/issues/160).

**Comment by avan36** [comment]
I have worked more on the noncompliance code and am hoping to finish it up tonight with further testing. I just have a question about making the noncompliance checks follow the logic from Kate’s paper. 

The paper says that “the following must be true,” but I’m not sure whether this means all the checks must pass (e.g., 3rd character US privacy string must be Y, OptanonConsent cookie contains isGpcEnabled=1). I think this interpretation is not the best, as if all 3 have to be true, 100% of sites became noncompliant. Even if we say only one check has to pass, then it becomes a problem because what if one check passes but another actively fails.

The code currently marks as noncompliant if any of the checks actively fail, not if any one passes – although we can switch this easily if preferred. I made strict versions of this interpretation and lenient versions. The strict version marks the site as noncompliant even if the string is null before and after. The lenient version says the site is compliant only if the site actively fails the check (meaning having a null value before & after will not cause it to fail, but having the string say not opted out before & after will). The strict version marks 70% of sites as noncompliant, while the lenient one is around 14%. I have kept both versions in the code – the strict version is currently commented out, to focus on sites that actively fail rather than ones with null values, but we can change this too.

I will work more on refining the logic for this tonight and checking that the code is accurate.

**Comment by SebastianZimmeck** [comment]
That is great, @avan36! 

> The paper says that “the following must be true,” but I’m not sure whether this means all the checks must pass (e.g., 3rd character US privacy string must be Y, OptanonConsent cookie contains isGpcEnabled=1).

To answer this question we need to say which sites we are considering. Per the [paper](https://sebastianzimmeck.de/hausladenEtAlGPCWeb2025.pdf):

> We consider a privacy string to be implemented if it has a non-null value. 

> Some sites had a non-null privacy string value either only before or only after sending a GPC signal.

So, for example, per Figure 9, n = 1,560 for December 2023 for the OptanonConsent cookie, the bars for before and after GPC show different totals because some sites only implement a string before or after but not both. So, my interpretation of that is that we consider a site to properly opt out if it has a null value before GPC and a non-null value set to opt out after GPC. On the other hand, we also consider a site to not opt out if it has a non-null non-opt out value before GPC and a null value after GPC. 

**So, the bottom line seems to me that if a string is not set to opt out after GPC, we have a non-compliant site.**

I believe @katehausladen used the Processing_Analysis_Data.ipynb in the [Google Drive](https://drive.google.com/drive/folders/1rmPkSw-X4KRx45X7PWAmnrF117h8Fdr-?usp=sharing) to create the graphs in the paper from the crawl data. So, can you check the logic, @avan36, if it is true what I am saying? If you run your logic on the data @katehausladen ran, you should come out at the same result, e.g., : 

> Compared to the other privacy strings, it has the highest percentages of being in an opted out state after a GPC signal; 66% (1,035/1,560) ...

As to:

> The code currently marks as noncompliant if any of the checks actively fail ...

if a site implements multiple privacy strings, it is non-compliant if one fails. So, let us flag that site (and the strings for which they fail). That is also what we did in the paper:

> In December 2023, 44% (1,411/3,226) of such sites opted users out via all implemented privacy strings

**Comment by Mattm27** [comment]
After updating the logic in our non-compliance detection [Colab](https://colab.research.google.com/drive/1HPV05RObE-J9ZvYgCqABNMY3zKhwK16m#scrollTo=GZ8o1Vx8ExOY) to replicate closer to the logic implemented in the Processing_Analysis_Data.ipyb which is used to generate figures, the percentage of non-compliant sites in the Feb/Mar 2025 CT crawl increased slightly from 14.06% to 14.96%. We observed similar increases in the CA and CO crawls for the same period. As privacy laws have become more prevalent over the past year and a half, it makes sense that compliance has gradually increased over time from what was reported in the Dec 2023 - April 2024 data.

I also did some manual spot checks—sampling about 15 sites for every 1,000 from the original crawl data that appeared non-compliant—and confirmed that each was correctly flagged as non-compliant in the Colab output.

I wasn’t able to locate the exact subsets of data that @katehausladen used to generate the figures in the paper for further cross-validation. Since the Colab we put together reports the overall percentage of non-compliant sites from the full batch of ~11,000 crawled domains (based on failing any one of the privacy string implementation checks), I’m wondering if we have rough historical comparisons from past full crawls—e.g., Dec 2023, Feb 2024, or April 2024? The paper appears to focus on compliance rates for individual privacy strings using a smaller sample, so a baseline from full crawls might be more useful for context.

**Comment by SebastianZimmeck** [comment]
Thank you, @Mattm27! 

> the percentage of non-compliant sites in the Feb/Mar 2025 CT crawl increased slightly from 14.06% to 14.96%. We observed similar increases in the CA and CO crawls for the same period. As privacy laws have become more prevalent over the past year and a half, it makes sense that compliance has gradually increased 

Do you mean the following?

> the percentage of compliant sites

On the following:

>  The paper appears to focus on compliance rates for individual privacy strings using a smaller sample

Yes, ideally we want to do the same. E.g., if there are 2,000 sites implementing the `isGpcEnabled` and 1,800 are changing their value to opt out, 1,800/2,000 is a more meaningful compliance rate than 1,800/<total crawled sites>. We would not expect a site to be compliant for technologies it does not implement.

**Comment by Mattm27** [comment]
As discussed in this morning’s meeting, the reported 14.96% reflects the percentage of sites out of the full ~11,000-site crawl that were deemed non-compliant. @avan36 and I will continue investigating how the figures in the paper were generated and attempt to recreate them using our own logic. This includes limiting the analysis to only those sites that implement at least one privacy string—such as the US Privacy String, the OptanonConsent cookie, or a GPP string—when calculating the final non-compliance rate.

As a side note, while using some older Colabs to process and create figures for more recent crawl data, I encountered an issue importing the cmp_api package for decoding GPP strings, due to recent changes in the repo’s file structure. Since this package is used across multiple Colabs, including for parsing raw crawl data, we’ll need to update these code chunks ahead of the next crawl.

**Comment by SebastianZimmeck** [comment]
> As discussed in this morning’s meeting, the reported 14.96% reflects the percentage of sites out of the full ~11,000-site crawl that were deemed non-compliant.

Thanks for clarifying that! I misunderstood this point [above](https://github.com/privacy-tech-lab/gpc-web-crawler/issues/136#issuecomment-2830561114).

> As a side note, while using some older Colabs to process and create figures for more recent crawl data, I encountered an issue importing the cmp_api package for decoding GPP strings, due to recent changes in the repo’s file structure.

Yes, it would be good if you can update that. The cmp_api now lives in the repo. Not sure if you can work with that directly. Maybe, you need some instructions like "download the cmp_api from the repo and put it in the Google Drive in the directory of the Colab" or something similar.

**Comment by Mattm27** [comment]
We resolved the issue by ensuring the correct path to the cmp_api_python directory was added to sys.path and importing the cmp_api package explicitly as a submodule (e.g., from cmp_api import cmpapi_test) in each Colab. To avoid conflicts from previous clones, we also added a step to remove any existing gpc-web-crawler directory before recloning the repo. With this fix in place, we can now move on to reproducing the graphs and further validating our non-compliance results—I'll be working on that tonight.

**Comment by SebastianZimmeck** [comment]
> We resolved the issue by ensuring the correct path to the cmp_api_python directory was added to sys.path and importing the cmp_api package explicitly as a submodule (e.g., from cmp_api import cmpapi_test) in each Colab.

Great! Can you clarify, the cmp_api package can live here on GitHub and does not need to be copied into the Google Drive? Or do you mean the cmp_api package is imported from the local copy of the gpc-web-crawler repo?

**Comment by Mattm27** [comment]
Yes! The cmp_api package is just imported from a local copy of the gpc-web-crawler repo, so the package can live here on github!

**Comment by Mattm27** [comment]
We encountered issues due to changes in the sheet structures and file names within the `Ad_Analysis_2023` folder in Google Drive. These changes affected several functions in [Monthly Data Analysis](https://colab.research.google.com/drive/17lW63ZBBFwozuMQ3TmzVXQNZFwGQwmot?usp=drive_link#scrollTo=3LPFdYEGetSt) Colab, but we were able to easily update those functions to restore functionality. Separately, the `apply_eval` function in [Processing Analysis Data](https://colab.research.google.com/drive/12e87xF7nPhIRE-JR4zAcNaWR88Yutv9Q#scrollTo=78M9-jNdszKk) Colab broke due to unexpected data formats caused by added columns. We implemented a manual workaround to temporarily patch the issue and generate the figures, but a permanent fix to the `apply_eval` function is probably needed.

Our [Non-Compliance Check](https://colab.research.google.com/drive/1HPV05RObE-J9ZvYgCqABNMY3zKhwK16m#scrollTo=9oggrkK6cUti) Colab reported that roughly ~15% (1,758 sites) of sites in the CT Feb/Mar 2025 crawl were not opting users out properly and are considered non-compliant, but this figure is based on all ~11,000 sites, including both successfully and unsuccessfully added ones. In contrast, the 21% (1,931 sites) non-complaint sites in CT Feb/Mar 2025 from the Monthly Data Analysis Colab reported in the attached figure reflects only the subset of sites  that were successfully added (9,081 sites). While the number of non-compliant sites should be the same, the discrepancy may in percentage may be due to the difference in sample size. 

<img width="640" alt="Image" src="https://github.com/user-attachments/assets/2cf1fa19-6aea-49fe-8d74-534f2d99860f" />

**Comment by SebastianZimmeck** [comment]
As discussed today, the plan is to:

1. Recreate figures in the paper, e.g., Figure 11 is relatively simple, with the same data as in the paper. We should come out at the exact same figure as @katehausladen. We do this for all the compliance figures (5, 6, 9, 11, 12, 13). 
2. Then, we can use new data and create the figures for those.
3. Then, we can adapt and develop a compliance dashboard (which, could be simple first, e.g., just displaying a list of non-compliant sites that @Mattm27 and @avan36 already started developing logic for; just copy-pasting the figures we generate).

**Comment by avan36** [comment]
Hello, apologies for the delay in an update. Kate’s monthly data analysis is working with modifications, although it has some slight quirks. I’ve [attached a doc](https://docs.google.com/document/d/11Ah_MzTMOHo6JQmf75k0_Mk3CnKEwpsLoclsUowFqbc/edit?usp=sharing) saying how our graphs compare to Kate’s with the same data as well (differences in the Optanon Consent Cookie Opt-Out n=1,558 vs n=1,560 and Tranco rank graph distribution, but everything else is the same). Now we’re trying to match Kate’s new working Colab to our noncompliance one to match them (it's hard to compare because our noncompliance Colab just verifies any check doesn't work and then just marks as noncompliant for the rest of the checks, as Kate's thesis checks before and after for each one regardless of whether the others passed). Once we have built up a more Kate-like analysis of what passed before and after, we will be able to compare our logic more carefully with her work (since we're relying on the same data). We have a basic tally at the moment, but it does not 

<img width="610" alt="Image" src="https://github.com/user-attachments/assets/f2136ae6-1ed1-4370-a5e1-3a8e1f470890" />

For the future: Since the monthly data analysis Colab doesn't give us an analysis on a case-by-case analysis, we will use our noncompliance Colab for this and Kate's Colab for an overall analysis.

**Comment by SebastianZimmeck** [comment]
Thank you, @avan36! That is good progress! 

> Since the monthly data analysis Colab doesn't give us an analysis on a case-by-case analysis, we will use our noncompliance Colab for this and Kate's Colab for an overall analysis.

Yes, we will use @katehausladen's Colab to create the summary graphs and @avan36's and @Mattm27's Colab to identify non-compliance for individual sites. Both Colabs are ingesting the same Google Sheet crawl data.

**Comment by avan36** [comment]
Matt and I have worked since our last meeting on making the percentages at the bottom of the Noncompliance Google Colab more similar to Kate's Monthly_Data_Analysis, so that it looks only at instances where that check (e.g, usp_api) was actually present, rather than blindly reporting on total noncompliance. Some numbers are very close to Kate's (within 2%) while others are less close (within 5%). I have been working to try to reduce this discrepancy, but more work is still needed – it's difficult, as both Processing_Analysis_Data and Monthly_Data_Analysis are so complex. I have also worked on making the code simpler/more explained for Ruby and Sage to take over. We are planning to meet tomorrow.

I have attached a document with comparisons.



<img width="504" alt="Image" src="https://github.com/user-attachments/assets/d4e8f6ea-71c1-464d-9f32-3ddda82dda28" />

**Comment by SebastianZimmeck** [comment]
> so that it looks only at instances where that check (e.g, usp_api) was actually present

Great! Thanks, @avan36! 

> Some numbers are very close to Kate's (within 2%) while others are less close (within 5%). 

When you say it is close (or not), I am assuming you mean the granular analysis and not the summary analysis, right? 

By granular analysis I mean creating statistics from comparing site A's US Privacy String, for example, before and after receiving a GPC signal as opposed to summary data that identifies, e.g., how many sites changed their US Privacy String in total vs how many did not (i.e., the recreation of @katehausladen's statistics from the paper). The latter should be exactly the same (given the same data).

**Comment by avan36** [comment]
Hi Sebastian, yes I mean the granular analysis, not the summary analysis!

**Comment by SebastianZimmeck** [comment]
@sagealtman and @RubyFri will be in the lead on this issue.

**Comment by SebastianZimmeck** [comment]
> In order for a site to be GPC-compliant, the following must be true ...

That is a good start! 

There are a number of edge cases we also need to consider:

- For example, is a site compliant if it does not have any GPP string at all after receiving the GPC signal (while it had one before)? What if the site has a GPP string after receiving the GPC signal but now the particular section is missing? 
- What if the national section of the GPP string contradicts the California section (assuming we are testing for California)?
- How do we handle multiple string contradictions? For example, the site is compliant as the US Privacy String is concerned but not with regard to GPP.
 
The reasons columns may be the solution for the last two cases as we do not necessarily need to make a call one way or the other but just surface the inconsistency.

> The current version of this collab results in the same statistics as Ambrose EXCEPT for the OptAnonConsentCompliant which has 421/1693 not passing (24.9%) which is much closer than the previous 30.6% to Kates 26.39%. 

That is progress. Though, we would need to get the exact same results. When using the same data and performing the same analysis, there is no reason why we are getting different results.

We can talk more about that tomorrow ...

**Comment by RubyFri** [comment]
Now that we have recreated the figures and numbers from Kate's collab and the paper, I am back at creating a script to identify non compliant sites. A non compliant site is determined by the following 4 items: US Privacy String, the GPP String, OptAnonConsent cookies and the wellknown/gpc.json.
The script I am currently working on uses the following rules to determine if a site is non compliant:

1. USPS strings are checked by their values in uspapi_before_gpc, uspapi_after_gpc, usp_cookies_before_gpc, and usp_cookies_after_gpc. If the before value is non-null, the after value must be non-null and have a Y in the position of the third character. If the before value is null, the after value can either be null or have a Y in the position of the third character. We treat the dash in the third character as a null value. All strings must have length 4. Some examples of compliant and noncompliant before and after privacy strings are as follows:

      Compliant or not enough information to mark as non compliant:
      - 1YNN to 1YYN
      - 1Y-N to 1Y-N
      - null to null
      - 1--- to 1---
      - null to 1-YN
  
     Non Compliant sites
      - 1--- to 1-N-
      - null to 1NNN
      - 1---N to 1---N (note the improper length here)
      - 1NYN to 1YNY
      - 1YNN to 1YNN
     
2. GPP Strings are checked by their decoded values of SaleOptOut and SharingOptOut (and TargetedAdvertisingOptOut if the GPP string is not a uscav1 string). We use the values from decoded_gpp_before_gpc and decoded_gpp_after_gpc to determine non compliance. 0 represents not applicable, 1 represents opted out, 2 represents did not opt out. A value is compliant if and only if it is decoded after with a value of 1, or goes from null to null (or no value to no value). Thus 0 to 0, 2 to 2, 0 to 2, or anything that is not something to 1 or null to null is marked non compliant.

3. OptAnonConsent cookie values are taken from OptanonConsent_before_gpc and OptanonConsent_after_gpc in the google sheets. For this, the possible values are either have isGPCEnabled=1 or 0, null, or “no_gpc”. Here I have a clarification question for @SebastianZimmeck : should we treat a value of no_gpc as compliant or noncompliant? I specifically have questions on the following case scenarios: 
1. if we have before and after both as no_gpc, is that equivalent to null to null where we don’t list it as non compliant due to too little information, or is that non compliant?
2. If it goes from null to no_gpc, is that non compliant or too little information?
3. From no_gpc to isGPCEnabled=1, just confirming that this would be compliant
Other than that I plan on marking anything that doesn't go from null to null or something to isGPCEnabled=1 as non compliant.

4. Finally we check the wellknown/gpc.json from the Well-known column in the google sheets. Here anything with "gpc": false will be marked as noncompliant.

**Comment by SebastianZimmeck** [comment]
> 1. USPS strings ... All strings must have length 4. ... not enough information to mark as non compliant:

What if they do not have length 4? Is "not enough information to mark as compliant" non-compliant or a third category? How did we categorize "incorrect format" strings in the paper?

> ... null to null ...

That is the "not applicable" category, right? So, we may have four categories:

1. Compliant
2. Not compliant
3. Not applicable
4. Incorrect format

> 2. GPP Strings ... goes from null to null ... or anything that is not something to 1 or null to null is marked non compliant.

There may be an "incorrect format" category here as well for GPP Strings (and OptanonConsent and .well-known). For example, `SaleOptOut = 8` or `SellingOptOut = 1`.

"null to null" is probably better "not applicable."

> OptAnonConsent cookie ... 

> Here I have a clarification question for @SebastianZimmeck : should we treat a value of no_gpc as compliant or noncompliant? I specifically have questions on the following case scenarios: ...

What is the meaning of no_gpc per OneTrust's specification?

**Comment by RubyFri** [comment]
Brief update on what I have been working on. The previous logic in the collab determining if certain GPP strings were non-compliant did not work as intended. I have since updated the code, so that instead of using the decoded gpp strings, it now takes in each websites GPP string (before decoding), then decodes it using Kate's python version of the iabgpp decoder into a valid python dictionary. I then use the resulting python dictionary to determine compliance following the updated [GPP Specification](https://docs.google.com/spreadsheets/d/1cB-Y2W46tn7orDs7EcZ7neWzjVYEpIQHjCw_p8YSa0Y/edit?usp=sharing) (ex: ShareOptOut must be 1 after GPC signal is sent to be compliant). 

To use Kate's python version I had to copy her cmp_api_python into the google drive [Processing_Data_Analysis](https://drive.google.com/drive/folders/1rmPkSw-X4KRx45X7PWAmnrF117h8Fdr-?usp=share_link) folder, and make a couple edits so that it would work as intended with my[ non compliance collab. ](https://colab.research.google.com/drive/184MxRTgsQ-9OkJKmP422drHs_aSsgboX?usp=sharing). However, her API threw exceptions and printed warnings if the GPP string was not formatted correctly or able to be decoded– this lack of a return value in some cases made my code logic more difficult in the collab, so I slightly edited her code so that if the string was ever unable to be decoded it returned an empty python dictionary {}. This simplified my logic when processing the decoded GPP dictionary for compliance. That said, my slightly edited version of the [cmp_api_python](https://drive.google.com/drive/folders/1TNS-_jtr8LWP5iPf6C5Hhng7aCKFoL7c?usp=share_link) is different from the version that appears in the GitHub. The version in the GitHub is the correct version and does not need to be edited, the only reason another slightly different version exists in the google drive is to simplify code logic locally when determining non compliant sites. I have mentioned in the readme that the version in the Processing_Data_Analysis folder is for local use only.

This new logic considers all RELEVANT sections of the GPP string, for example, it looks at a websites usnatv1 section and its uscov1 section if and only if the site is from the CO crawl. If a website has a uscav1 (CA section) this section will only be checked for non compliance if the crawl was done in California. The usnatv1 sections are always checked for compliance in their SaleOptOut, ShareOptOut, and TargetedAdvertisingOptOut categories. 

I am still looking into the GPC subsection of the GPP string and if its value has anything to do with a website being considered noncompliant. So far, it seems not according to the [GPP Specification](https://github.com/InteractiveAdvertisingBureau/Global-Privacy-Platform/blob/main/Sections/US-National/IAB%20Privacy’s%20Multi-State%20Privacy%20Agreement%20(MSPA)%20US%20National%20Technical%20Specification.md) but I will continue to look and will update this issue as I gain more clarity on the GPC subsection.

**Comment by SebastianZimmeck** [comment]
I will look at @RubyFri's [Classification of Compliance](https://docs.google.com/spreadsheets/d/1cB-Y2W46tn7orDs7EcZ7neWzjVYEpIQHjCw_p8YSa0Y/edit?usp=sharing).

**Comment by SebastianZimmeck** [comment]
Thanks @RubyFri for your work on the Classifications_of_Compliance Sheet. Can you fix the following points?

- Substantive Points
    - Per [above comment](https://github.com/privacy-tech-lab/gpc-web-crawler/issues/136#issuecomment-3049936086), can you add how the case is handled if there is a national section GPP string before and state section GPP after? I see your comment "GPP String are only looked at in their relevent state (uscav1, usctv1, uscov1) and/or the usnatv1 sting if there." But that does not define how such cases are handled.
    - One other important point (and I know we have discussed it to do like you did), I find it hard to digest all the USPS combinations. So, let us just use **** when it does not really matter. E.g., there can be any string values before (****) for the string after 1YYN. So, to get that point across it is better to not enumerate all the possibilities. We can explain in the legend what the * means. I started working on this. It would be good, also for your own clarity, if you continue and boil it down to the essence.
    - Here are the values you have:
       - USPS: `Not Applicable`, `Opted Out`, `Did Not Opt Out`
       - OptAnonConsent: `Not Applicable`, `Opted Out`, `Did Not Opt Out`, `Not Compliant`
       - Well-Known: `Not Applicable`, `Opted Out`, `Did Not Opt Out`, `Not Compliant`, `Invalid Format`
       - GPP Strings: `Not Applicable`, `Compliant`, `Not Compliant and MissingAfterGPPString`, `Not Compliant`, `Invalid Format`
       **_Overall, the systematization does not seem to be fully fleshed out. It would be good if you think some more about and revise it to make it more consistent and consider all cases that could happen for all strings._** A few details:
       - The difference between "Did Not Opt Out" and "Not Compliant" is not clear to me. It seems that either we draw the conclusion already "Not Compliant" or keep it to a fact "Did Not Opt Out." Either way works, but we should be consistent. I am assuming that every site in the compliance sheet is not compliant. So, I do not think that explicitly needs to be spelled out. But in any case the system could use some refinement or clarification. If we keep "Opt Out," it also seems more natural to me to either have "Did Opt Out" and "Did Not Opt Out" or "Opted Out" and "Not Opted Out" but not a mixture of both.
       - Some of the tags also do not seem to be used (e.g., "Invalid" in the Well-Known). Either we delete them or we have one set of tags across all tabs. The later is better in my view. Along this line, why is there no "Invalid" in USPS? 
       - Also, the "Not Compliant and MissingAfterGPPString" is one edge case specifically tagged. That is a bit odd. We can either have more specific compliance categories or subcategories or an explanation column. But I am not sure why this is treated specifically like this. There could be also a "MissingAfterUSPSString." But that is not there.
       - There are columns that do not seem to do anything, (e.g., in Well-Known. "Anything that is not None or null that does not contain a gpc section  will appear in the PoteniallyNonCompliant Sheet with the reason Invalid_WellKnown") Should there go anything in there or is this just information? In the latter case, I do not think we need this column.
 

- Formalities
    - Generally, I find the coloring of the dropdown choices a bit unintuitive, especially, red for "Not Applicable" and blue for "Not Compliant." Can you fix that and also have unique colors. Blue is used for multiple.
    - Please run a spell check ("relevent" vs "relevant," "sting" vs "string").
    - Can you use the terminology in the same way as we use it in the Crawl_Data_CA and other files? For example, instead of "USPS Before GPC Signal" let us use "usps_before_gpc." If we use the same terminology, it is clear that it means the same thing (it is clear in my example, but in other cases it may not be).
    - Let us use camel case wherever appropriate. For example, our tab naming convention would be "GPPStrings" (I would actually use "GPP" because we speak of "USPS" and "GPP" without "String" so there is no reason for the difference).
    - The GPC spec also has camel case, for example, `lastUpdate' instead of `last update'.
    - Let's change the NOTEs to a legend, e.g., move "NOTE: Anything marked as Did Not Opt Out will appear in the PoteniallyNonCompliant Sheet with the reason USP_API and/or USP_COOKIES" to a legend tab. Also, let's include the name of and link to the referenced Sheet. Otherwise, it is unclear for someone not in the know where this PoteniallyNonCompliant Sheet is.
    - "gpc": false (lower case; just everything spelled the way how it is spelled in wherever this is referenced.
    - Let us change the cookie name to its regular capitalization, "OptanonConsent."

**Comment by SebastianZimmeck** [comment]
Here is an [interesting Stay SaaSy post](https://blog.staysaasy.com/p/you-have-too-many-metrics) that discusses the broader context of metrics.

**Comment by SebastianZimmeck** [comment]
I have substantially revised our approach for classifying whether a site is compliant given its privacy string settings. 

The new [Compliance Classifications are in the same Google Sheet as before](https://docs.google.com/spreadsheets/d/1cB-Y2W46tn7orDs7EcZ7neWzjVYEpIQHjCw_p8YSa0Y/edit?usp=sharing). My goal was to make the approach more general, systematic, comparable across different privacy strings, and easier to implement in our [GPC Web UI](https://github.com/privacy-tech-lab/gpc-web-ui). 

For example, in the previous approach we focused on non-compliance (for identifying enforcement targets), but we also want to show compliance (for identifying to which extent enforcement actions were successful). Thus, we are no longer talking about "reasons for non compliance" but "Compliance Classifications." This new approach is at the center of our analysis and fundamentally impacts the post-processing of crawl data and surfacing data in the [UI](https://github.com/privacy-tech-lab/gpc-web-ui/issues/26#issuecomment-3962298439) and [OptMeowt](https://github.com/privacy-tech-lab/gpc-optmeowt/issues/508).

A good starting point for understanding the new approach is to start with the legend I included in the first tab of the [Compliance Classifications](https://docs.google.com/spreadsheets/d/1cB-Y2W46tn7orDs7EcZ7neWzjVYEpIQHjCw_p8YSa0Y/edit?usp=sharing). We can go over it tomorrow in our meeting.

There are various concrete tasks as a consequence of this change. We need to:
- [x] **Implement the logic from the Compliance Classifications in the post-processing scripts** for identifying the (formerly) "LikelyNonCompliantSites." Generally, the logic should be much simpler than it used to be, especially, for the US Privacy String. There may still be some issues to adapt to go from the Compliance Classifications to program logic. There may still be some holes. We will find them as we go along. For example, I am not 100% sure whether the combinations of privacy strings I defined cover all combinations of privacy strings that we have in our data. I am also not sure how to best count the absolute numbers and percentages. These are questions that we should discuss as part of the implementation.
   - [Create per-site compliance classification format](https://github.com/privacy-tech-lab/gpc-web-crawler/issues/241#issuecomment-4002851739)
   - [Stop creating dedicated NullSite tabs](https://github.com/privacy-tech-lab/gpc-web-crawler/issues/237)
- [x] **Integrate the Compliance Classifications into the regular crawl results** as opposed to have a dedicated Sheet. The latter approach is an artifact from when we did not have a UI, and there was no easy way to quickly get a list of non-compliant sites. However, at this stage, it is better to have an additional column, "compliance_classifications", in each Crawl_Data_<State> tab. We can generate the Compliance Classifications with a dedicated Colab, as it is currently the case, or integrate the generation into the general post-processing Colab. Either way works. We can still populate the Compliance Classification column with the resulting labels as it currently is the case.
- [x] As we adopt the new approach, we would need to **re-run the "LikelyNonCompliantSites"** (i.e., now "Compliance Classification") for all states and all dates for which we have data. Once that is done, we can delete the NonCompliantSites (I dropped the "likely." Let us do that from now on.) tabs in the crawl data.
   - [Redo post-processing of California data](https://github.com/privacy-tech-lab/gpc-web-ui/issues/30) 
   - [Clarify relationship between accuracy results in crawl data and dedicated accuracy results](https://github.com/privacy-tech-lab/gpc-web-crawler/issues/236)
- [x] Redo post-processing due to [inconsistency fix of Compliance Classifications](https://github.com/privacy-tech-lab/gpc-web-crawler/issues/255)
- [x] We need to **update the [Wiki Instructions for Lab Members Performing Crawls](https://github.com/privacy-tech-lab/gpc-web-crawler/wiki/Instructions-for-Lab-Members-Performing-Crawls)** with the new method for generating the Compliance Classifications. For example, the following will no longer be true
   > Now, open the Non_Compliance_Colab and update the crawl_state and date variables. Run this colab to create a tab titled PotentiallyNonCompliantSites. To see the criteria of what makes a site potentially not compliant, open the Classifications_of_Compliance sheet.

   I already made the change to "Compliance Classifications" in this repo readme and Google Drive readme. The Compliance Classification Sheet is linked publicly in the repo readme.
  - https://github.com/privacy-tech-lab/gpc-web-crawler/issues/248
- [x] We would also need to update the Wiki and/or readme of the Web UI repo.
- [x] More tangentially related, check column headers in the crawl sheets and the legend as well as the descriptions of the headers in the readme of this repo to make sure they are still up to date and aligned (@SebastianZimmeck). 

Go over legend, Wiki, and readmes (in repo and Google drive) to make sure it is up to date
- [x] [Check that legend in crawl data Google Sheets and repo readme is up to date and fix as necessary](https://github.com/privacy-tech-lab/gpc-web-crawler/issues/245)

I created separate comments and issues for the [UI features](https://github.com/privacy-tech-lab/gpc-web-ui/issues/26#issuecomment-3962298439) and [null sites impacted](https://github.com/privacy-tech-lab/gpc-web-crawler/issues/237).

**Comment by SebastianZimmeck** [comment]
@mcrich921, can you clarify what [still needs doing](https://github.com/privacy-tech-lab/gpc-web-crawler/issues/136#issuecomment-3962336645)?

**Comment by SebastianZimmeck** [comment]
I believe the core work of this issue is done. There is more work related to the new Compliance Classification, which, however, we will address in individual issues (e.g., privacy-tech-lab/gpc-web-crawler#216, privacy-tech-lab/gpc-web-ui#92, and [issue 26 in the Web UI repo](https://github.com/privacy-tech-lab/gpc-web-ui/issues/26)).
